# Cython & C++: Why It Is Worth It

* Author: Cody W. Eilar <cody.eilar@gmail.com> 
* Course: ECE 633
* Professor: Dr. Marios Pattichis

In this notebook, I demo why I spent the time to write Cython and C++ code. 

In [1]:
# import custom modules
import py_video_segment as pvs
import numpy as np

In [2]:
# Define python function to get the number of false positives
def PyNumberFalsePositives(truth, predicted):
    count = 0; 
    for i in predicted: 
        if not((i in truth)): 
            count = count + 1; 
    return count

truth = np.array([1, 2, 3, 9, 10, 11]); 
predicted = np.array([4, 5, 6, 7, 8]); 

length = 100; 



In [3]:
%timeit PyNumberFalsePositives(truth, predicted)

The slowest run took 6.69 times longer than the fastest. This could mean that an intermediate result is being cached 
100000 loops, best of 3: 15.6 µs per loop


In [4]:
ma = pvs.PyMethodAccuracy(truth, predicted, 10)
%timeit ma.NumFalsePositives()

Created PyMethodAccuracy
Size is : 6
Size is : 5
The slowest run took 10.67 times longer than the fastest. This could mean that an intermediate result is being cached 
1000000 loops, best of 3: 464 ns per loop


## Analysis
As you can see, we are running *significantly* faster with the compiled code. 464 ns vs 15.6 micro seconds per loop. That is a 33x speed up for the same algorithm. Thus it is key to implement bottleneck code at the C++/Cython level. 